Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm

```diff
@@ -0,0 +1,299 @@
+    return ensureDict(maybeSidebarAction)
+        .and_then([](RetainPtr<NSDictionary<NSString *, id>> sidebarAction) { return ensureStringValue(sidebarAction.get(), sidebarActionTitleKey); })
+        .value_or(extension.displayName());
+}
+
+static std::optional<String> getDefaultSidebarPathFromExtension(WebExtension& extension)
+{
+    NSDictionary *manifest = extension.manifest();
+    auto maybePanelTitle = ensureDict(manifest[sidebarActionManifestKey])
+        .and_then([](RetainPtr<NSDictionary<NSString *, id>> sidebarAction) {
+            return ensureStringValue(sidebarAction.get(), sidebarActionPathKey);
+        });
+
+    // since std::optional::or_else cannot change the contained type, we have to early return here rather than
+    // continuing to use monadic operations
+    if (maybePanelTitle)
+        return maybePanelTitle;
+
+    return ensureDict(manifest[sidePanelManifestKey])
+        .and_then([](RetainPtr<NSDictionary<NSString *, id>> sidePanel) { return ensureStringValue(sidePanel.get(), sidePanelPathKey); })
+        .value_or(fallbackPath);
+}
+
+static std::optional<NSDictionary *> getDefaultIconsDictFromExtension(WebExtension& extensions)
+{
+    // FIXME: <https://webkit.org/b/276833> implement this
+    return std::nullopt;
+}
+
+template<typename T>
+static std::optional<T *> nilToNullopt(T *maybeNil)
+{
+    if (maybeNil)
+        return maybeNil;
+    return std::nullopt;
+}
+
+WebExtensionSidebar::WebExtensionSidebar(WebExtensionContext& context, IsDefaultSidebar isDefault) : WebExtensionSidebar(context, std::nullopt, std::nullopt, isDefault) { };
+
+WebExtensionSidebar::WebExtensionSidebar(WebExtensionContext& context, WebExtensionTab& tab) : WebExtensionSidebar(context, tab, std::nullopt, IsDefaultSidebar::No) { };
+
+WebExtensionSidebar::WebExtensionSidebar(WebExtensionContext& context, WebExtensionWindow& window) : WebExtensionSidebar(context, std::nullopt, window, IsDefaultSidebar::No) { };
+
+WebExtensionSidebar::WebExtensionSidebar(WebExtensionContext& context, std::optional<Ref<WebExtensionTab>> tab, std::optional<Ref<WebExtensionWindow>> window, IsDefaultSidebar isDefault)
+    // for this API, we never want to override both window and tab -- can only do one at a time
+    // if we somehow (erroneously) get both, prefer tab (most specific)
+    : m_context(WeakRef(context)), m_tab(tab), m_window(window && tab ? std::nullopt : window), m_isDefault(isDefault)
+{
+
+    // if this is the default action, initialize with default sidebar path / title if present
+    if (isDefaultSidebar()) {
+        auto& extension = context.extension();
+        m_titleOverride = getDefaultSidebarTitleFromExtension(extension);
+        m_sidebarPathOverride = getDefaultSidebarPathFromExtension(extension);
+        m_iconsOverride = getDefaultIconsDictFromExtension(extension);
+    }
+}
+
+std::optional<Ref<WebExtensionContext const>> WebExtensionSidebar::extensionContext() const
```

Comment:
We don't get much from this being `optional` as you only use `value().get()` when it could be `RefPtr` instead.

Reply:
There are a couple places where I'm using some of the monadic optional operations on `extensionContext` -- most notably in `canProgrammaticallyOpenSidebar` and `canProgrammaticallyCloseSidebar`. When I address the FIXMEs in both of those methods, I think I will also add another chained call to `and_then` in each to check that the delegates respond to the appropriate selectors, so I think the final implementation might be cleaner if I retain the `optional`s -- would it be alright to keep this one as an `optional<Ref<..>>`?